### PR TITLE
Hadoop FileSystemContract Conformance

### PR DESCRIPTION
**Pull request by robert-schmidtke**
This PR adds adherence to the Hadoop FileSystemContract for the XtreemFS Hadoop adapter. This is ensured by executing the FileSystemContractBaseTest with XtreemFS as the underlying file system.

Notable changes:
* .travis.yml adds the travis-contrib test suite which currently tests only the XtreemFS Hadoop adapter
* contrib/hadoop/build.xml adds targets which set the classpath for compiling and executing the JUnit tests from the command line. It is necessary to have HADOOP_PREFIX pointing to a local Hadoop installation for this. In NetBeans, add libraries 'hadoop_common_tests' (pointing to the $HADOOP_PREFIX/share/hadoop/common/hadoop-common-x.y.z-tests.jar file) and 'hadoop_common_dependencies' (pointing to the $HADOOP_PREFIX/share/hadoop/common/lib/*.jar files) in case of broken reference errors.
* contrib/hadoop/nbproject/build-impl.xml reflects changes in the project structure (mainly the division into src/main and src/test folders and the addition of JUnit tests) as well as some NetBeans specifics (I used NetBeans 8.0.2).
* contrib/hadoop/nbproject/project.properties cleans up the classpaths a bit and adds the new JUnit dependencies.
* contrib/hadoop/src/main/.../XtreemFSFileSystem.java:
  - uses setConf/getConf instead of holding a private instance,
  - does not depend on the fs.defaultFS property to be pointing to an XtreemFS volume, as this does not have to be the case (might as well be hdfs://...) and as a result de-deprecates the 'xtreemfs.defaultVolumeName' property (but it may fall back to fs.defaultFS),
  - returns the scheme properly without '://',
  - throws an IOException when not explicitly overwriting an existing file upon creation,
  - applies the umask when set in the configuration via CommonConfigurationKeys.FS_PERMISSIONS_UMASK_KEY when creating files and directories,
  - does not overwrite existing files during 'rename' if not configured to do so via 'xtreemfs.rename.overwrite' (this is Hadoop specific and not POSIX),
  - allows renaming a file to a directory resulting in the file being moved there instead of having the MoveOperation throw an exception,
  - throws an exception when trying to non-recursively delete a non-empty directory instead of just returning false,
  - throws an exception when obtaining the status of a non existing file instead of returning null,
  - always qualifies the working directory,
  - throws an exception when attempting to create a subdirectory of a file instead of returning false.
* contrib/hadoop/src/test/.../XtreemFSFileSystemContractTest.java implements the FileSystemContractBaseTest provided by Hadoop and initializes it with XtreemFS.
* doc/xtreemfs-userguide/xtfs-guide.tex explains the additional option available in the XtreemFS Hadoop adapter to control overwriting of existing files during rename
* tests/test_config.py fixes some syntax errors and adds the Hadoop JUnit tests to the 'full' and 'travis-contrib' test suites.
* tests/test_scripts/hadoop_junit_tests.sh executes the Hadoop JUnit tests for specific Hadoop versions, currently only 2.7.1.

### CODE REVIEW

**Comment by robert-schmidtke** [comment]
Now that the unit tests are passing I'd be happy about a code review, @chrkl

File: contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java [hunk 1]
```diff
@@ -84,26 +84,14 @@ public void initialize(URI uri, Configuration conf) throws IOException {
             Logging.logMessage(Logging.LEVEL_DEBUG, this, "init : " + uri);
         }
 
-        String defaultURIString = conf.get("fs.defaultFS");
```
Comment (chrkl) [inline contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java]: As I remember correctly, we introduced fs.defaultFS as this seems to be the typical way to point to a default file system. Maybe @jan--f can clarify this. I would prefer to allow both options for compatibility purposes.

**Comment by chrkl** [comment]
Good work. I agree with your changes, except the defaultFS thing.

File: contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java (hunk 1 above)
Comment (robert-schmidtke) [inline contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java]: It's just that XtreemFS may not always be the default file system, e.g. when you mix sources from HDFS and XtreemFS. In that case, querying fs.defaultFS will give the HDFS URI. I will add a fallback to the URI passed to the constructor (typically the file URI through which XtreemFS is accessed the first time), and if that does not contain the volume then I'll add another fallback to fs.defaultFS.

Comment (chrkl) [inline contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java]: Ok, agreed!

Comment (jan--f) [inline contrib/hadoop/src/main/java/org/xtreemfs/common/clients/hadoop/XtreemFSFileSystem.java]: yes I agree too. Fallback code is a good idea. never thought about mixing sources before.

**Pull request merged**